File: typescript/src/swagger20-engine.ts
import {
  SWAGGER20_METHODS,
  Swagger20Document,
  isSwagger20Object,
  objectMember,
  operationInfo,
  pathItemMemberResource,
  type Swagger20LoadOptions,
  type Swagger20Method,
  type Swagger20OperationInfo,
  type Swagger20ResolvedOperation,
  type Swagger20Source,
} from "./swagger20-model.js";
import { loadSwagger20Document } from "./swagger20-loader.js";
import { decodePointerToken, escapePointerToken, newSwagger20ResolutionMemo, wellFormedPointerToken } from "./swagger20-reference.js";

export interface Swagger20PrepareOptions extends Swagger20LoadOptions {
  source: Swagger20Source;
  ref: string;
  /** OpenAPI-native execution configuration; later passes fill this surface. */
  context?: Record<string, unknown>;
}

/** A failure produced by the exact Swagger 2.0 engine lane. */
export class Swagger20ExecutionError extends Error {
  readonly code: string;
  readonly details?: unknown;
  readonly evidence?: unknown;

  constructor(code: string, message: string, options: { cause?: unknown; details?: unknown; evidence?: unknown } = {}) {
    super(message, options.cause === undefined ? undefined : { cause: options.cause });
    this.name = "Swagger20ExecutionError";
    this.code = code;
    if (options.details !== undefined) this.details = options.details;
    if (options.evidence !== undefined) this.evidence = options.evidence;
  }
}

/** One selected operation in the edition-specific native model. */
export class PreparedSwagger20Operation {
  readonly document: Swagger20Document;
  readonly operation: Swagger20ResolvedOperation;
  readonly info: Swagger20OperationInfo;
  /** @internal */ readonly options: Swagger20PrepareOptions;

  /** @internal */
  constructor(document: Swagger20Document, operation: Swagger20ResolvedOperation, info: Swagger20OperationInfo, options: Swagger20PrepareOptions) {
    this.document = document;
    this.operation = operation;
    this.info = info;
    this.options = options;
  }

  /** Pass-one boundary: later edition files add native request execution. */
  async execute(_input?: unknown): Promise<{ outputPresent: boolean; output?: unknown }> {
    throw new Swagger20ExecutionError("ERR_REFUSED", "Swagger 2.0 operation execution requires request-surface preparation");
  }
}

/** Loads and selects one operation through only the exact Swagger 2.0 lane. */
export async function prepareSwagger20(options: Swagger20PrepareOptions): Promise<PreparedSwagger20Operation> {
  let document: Swagger20Document;
  try {
    document = await loadSwagger20Document(options.source, options);
  } catch (error: unknown) {
    throw new Swagger20ExecutionError("SOURCE_LOAD_FAILED", errorMessage(error), { cause: error });
  }
  const operation = await resolveSwagger20Operation(document, options.ref);
  return new PreparedSwagger20Operation(document, operation, operationInfo(operation, options.ref), options);
}

/** Exact selector validation without artifact loading. */
export function validateSwagger20Selector(ref: string): void {
  parseSwagger20Ref(ref);
}

/** Enumerates every resolvable fixed Swagger 2.0 operation slot. */
export async function listSwagger20Operations(document: Swagger20Document): Promise<Swagger20OperationInfo[]> {
  const paths = objectMember(document.root, "paths");
  if (!paths.valid) return [];
  const result: Swagger20OperationInfo[] = [];
  for (const path of Object.keys(paths.value!).sort()) {
    for (const method of SWAGGER20_METHODS) {
      const ref = `#/paths/${escapePointerToken(path)}/${method}`;
      try {
        const operation = await resolveSwagger20Operation(document, ref);
        result.push(operationInfo(operation, ref));
      } catch { /* smallest-owner confinement omits only this target */ }
    }
  }
  return result;
}

/** @internal */
export async function resolveSwagger20Operation(
  document: Swagger20Document,
  ref: string,
): Promise<Swagger20ResolvedOperation> {
  let parsed: { path: string; method: Swagger20Method };
  try {
    parsed = parseSwagger20Ref(ref);
  } catch (error: unknown) {
    throw new Swagger20ExecutionError("INVALID_OPERATION_REF", errorMessage(error), { cause: error });
  }
  const paths = objectMember(document.root, "paths");
  if (!paths.present || !paths.valid) {
    throw new Swagger20ExecutionError("ERR_REFUSED", "Swagger 2.0 document has no usable paths object");
  }
  const rawItem = paths.value![parsed.path];
  if (!isSwagger20Object(rawItem)) {
    throw new Swagger20ExecutionError("OPERATION_NOT_FOUND", `operation ${JSON.stringify(ref)} was not found`);
  }
  let item;
  try {
    item = await document.graph.resolvePathItem({ raw: rawItem, resource: document.entry }, parsed.method, newSwagger20ResolutionMemo());
  } catch (error: unknown) {
    throw new Swagger20ExecutionError("ERR_REFUSED", errorMessage(error), { cause: error });
  }
  const rawOperation = item.raw[parsed.method];
  if (!isSwagger20Object(rawOperation)) {
    throw new Swagger20ExecutionError("OPERATION_NOT_FOUND", `operation ${JSON.stringify(ref)} was not found`);
  }
  return {
    raw: rawOperation,
    resource: pathItemMemberResource(item, parsed.method),
    pathItem: item,
    path: parsed.path,
    method: parsed.method,
  };
}

function parseSwagger20Ref(ref: string): { path: string; method: Swagger20Method } {
  const prefix = "#/paths/";
  if (!ref.startsWith(prefix)) throw new Error(`selector ${JSON.stringify(ref)} must have exact form #/paths/<escaped-path>/<lowercase-method>`);
  const parts = ref.slice(prefix.length).split("/");
  if (parts.length !== 2 || parts[0] === "") throw new Error(`selector ${JSON.stringify(ref)} must have exact form #/paths/<escaped-path>/<lowercase-method>`);
  if (!wellFormedPointerToken(parts[0]!)) throw new Error(`selector ${JSON.stringify(ref)} contains a malformed RFC 6901 path token`);
  const method = parts[1] as Swagger20Method;
  if (!SWAGGER20_METHODS.includes(method)) throw new Error(`selector ${JSON.stringify(ref)} has inadmissible lowercase Swagger 2.0 method ${JSON.stringify(parts[1])}`);
  return { path: decodePointerToken(parts[0]!), method };
}

function errorMessage(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: go/swagger20_engine.go
package openapiclient

import (
	"context"
	"fmt"
	"net/http"
	"sort"
	"strings"
	"sync"
)

// Swagger20PrepareOptions is the edition-specific preparation surface. It is
// deliberately separate from PrepareOptions so Source.Document remains typed
// as *openapi3.T and the two edition lanes never share a structural model.
type Swagger20PrepareOptions struct {
	Source                 Swagger20Source
	Ref                    string
	Context                map[string]any
	HTTPClient             *http.Client
	Server                 string
	RequestMedia           string
	PropertyMedia          map[string]string
	ParameterConverter     ParameterConverter
	EmptyValueForm         Swagger20EmptyValueForm
	RequestContentCodings  map[string]ContentEncoder
	ResponseContentCodings map[string]ContentDecoder
	MaxDeliveryUnitBytes   int64
	AllowExternalRefs      *bool
}

type Swagger20PreparedOperation struct {
	document      *Swagger20Document
	operation     swagger20Operation
	info          OperationInfo
	options       Swagger20PrepareOptions
	parameterOnce sync.Once
	parameterSet  *swagger20ParameterSet
	parameterErr  error
}

func (p *Swagger20PreparedOperation) Ref() string {
	if p == nil {
		return ""
	}
	return p.info.Ref
}

func (p *Swagger20PreparedOperation) Info() OperationInfo {
	if p == nil {
		return OperationInfo{}
	}
	info := p.info
	info.Tags = append([]string(nil), info.Tags...)
	return info
}

// Parameters returns the effective Swagger 2.0 parameter identities after
// reference resolution, override, and declaration-only exclusion checks.
func (p *Swagger20PreparedOperation) Parameters() ([]Swagger20ParameterInfo, error) {
	parameters, err := p.parameters()
	if err != nil {
		return nil, err
	}
	result := make([]Swagger20ParameterInfo, len(parameters.all))
	for index, parameter := range parameters.all {
		result[index] = parameter.info()
	}
	return result, nil
}

func (p *Swagger20PreparedOperation) parameters() (*swagger20ParameterSet, error) {
	if p == nil {
		return nil, &ExecutionError{Code: CodeRefused, Message: "Swagger 2.0 prepared operation is nil"}
	}
	p.parameterOnce.Do(func() {
		p.parameterSet, p.parameterErr = effectiveSwagger20Parameters(p.document.graph, p.operation)
		if p.parameterErr != nil {
			p.parameterErr = &ExecutionError{Code: CodeRefused, Message: p.parameterErr.Error(), Cause: p.parameterErr}
		}
	})
	return p.parameterSet, p.parameterErr
}

// PrepareSwagger20 selects one operation through the exact Swagger 2.0 gate
// and JSON-Pointer selector grammar. It never invokes the OpenAPI 3.x loader.
func (e *Engine) PrepareSwagger20(ctx context.Context, options Swagger20PrepareOptions) (*Swagger20PreparedOperation, error) {
	requestCodings, err := normalizeContentEncoders(options.RequestContentCodings)
	if err != nil {
		return nil, &ExecutionError{Code: CodeRefused, Message: err.Error(), Cause: err}
	}
	responseCodings, err := normalizeContentDecoders(options.ResponseContentCodings)
	if err != nil {
		return nil, &ExecutionError{Code: CodeRefused, Message: err.Error(), Cause: err}
	}
	options.RequestContentCodings = requestCodings
	options.ResponseContentCodings = responseCodings
	loadClient := options.HTTPClient
	if loadClient == nil && e != nil {
		loadClient = e.client
	}
	if loadClient == nil {
		loadClient = defaultHTTPClient()
	}
	if options.HTTPClient == nil {
		if e != nil {
			options.HTTPClient = e.invocationClient
		} else {
			options.HTTPClient = defaultInvocationHTTPClient()
		}
	}
	allowExternal := true
	if options.AllowExternalRefs != nil {
		allowExternal = *options.AllowExternalRefs
	}
	document, err := loadSwagger20Document(ctx, loadClient, options.Source, allowExternal)
	if err != nil {
		return nil, &ExecutionError{Code: CodeSourceLoadFailed, Message: err.Error(), Cause: err}
	}
	operation, info, err := resolveSwagger20Operation(document, options.Ref)
	if err != nil {
		return nil, err
	}
	return &Swagger20PreparedOperation{document: document, operation: operation, info: info, options: options}, nil
}

func resolveSwagger20Operation(document *Swagger20Document, ref string) (swagger20Operation, OperationInfo, error) {
	path, method, err := parseSwagger20Ref(ref)
	if err != nil {
		return swagger20Operation{}, OperationInfo{}, &ExecutionError{Code: CodeInvalidRef, Message: err.Error(), Cause: err}
	}
	paths := swagger20DocumentObject{raw: document.root, resource: document.entry}.paths()
	if !paths.present || !paths.valid {
		return swagger20Operation{}, OperationInfo{}, &ExecutionError{Code: CodeRefused, Message: "Swagger 2.0 document has no usable paths object"}
	}
	rawItem, present := paths.value.member(path)
	if !present {
		return swagger20Operation{}, OperationInfo{}, &ExecutionError{Code: CodeRefNotFound, Message: fmt.Sprintf("operation %q was not found", ref)}
	}
	itemObject, ok := rawItem.(map[string]any)
	if !ok {
		return swagger20Operation{}, OperationInfo{}, &ExecutionError{Code: CodeRefNotFound, Message: fmt.Sprintf("operation %q was not found", ref)}
	}
	item := swagger20PathItem{raw: swagger20Object(itemObject), resource: document.entry}
	item, err = document.graph.resolvePathItem(item, method, newSwagger20ResolutionMemo())
	if err != nil {
		return swagger20Operation{}, OperationInfo{}, &ExecutionError{Code: CodeRefused, Message: err.Error(), Cause: err}
	}
	declared := item.operation(method)
	if !declared.present || !declared.valid {
		return swagger20Operation{}, OperationInfo{}, &ExecutionError{Code: CodeRefNotFound, Message: fmt.Sprintf("operation %q was not found", ref)}
	}
	operation := swagger20Operation{
		raw: declared.value, resource: item.resourceFor(method), pathItem: item, path: path, method: method,
	}
	return operation, operation.operationInfo(ref), nil
}

func parseSwagger20Ref(ref string) (path string, method string, err error) {
	const prefix = "#/paths/"
	if !strings.HasPrefix(ref, prefix) {
		return "", "", fmt.Errorf("selector %q must have exact form #/paths/<escaped-path>/<lowercase-method>", ref)
	}
	parts := strings.Split(ref[len(prefix):], "/")
	if len(parts) != 2 || parts[0] == "" {
		return "", "", fmt.Errorf("selector %q must have exact form #/paths/<escaped-path>/<lowercase-method>", ref)
	}
	if !wellFormedPointerToken(parts[0]) {
		return "", "", fmt.Errorf("selector %q contains a malformed RFC 6901 path token", ref)
	}
	method = parts[1]
	if !swagger20Method(method) {
		return "", "", fmt.Errorf("selector %q has inadmissible lowercase Swagger 2.0 method %q", ref, method)
	}
	path = decodePointerToken(parts[0])
	return path, method, nil
}

// ValidateSwagger20Selector validates the exact Swagger 2.0 operation
// selector spelling without loading or dereferencing an artifact.
func ValidateSwagger20Selector(selector string) error {
	_, _, err := parseSwagger20Ref(selector)
	return err
}

func (d *Swagger20Document) operations() []OperationInfo {
	if d == nil || d.graph == nil || d.entry == nil {
		return nil
	}
	paths := swagger20DocumentObject{raw: d.root, resource: d.entry}.paths()
	if !paths.present || !paths.valid {
		return nil
	}
	pathNames := make([]string, 0, len(paths.value))
	for path := range paths.value {
		pathNames = append(pathNames, path)
	}
	sort.Strings(pathNames)
	methods := []string{"get", "put", "post", "delete", "options", "head", "patch"}
	var result []OperationInfo
	for _, path := range pathNames {
		for _, method := range methods {
			ref := "#/paths/" + escapeJSONPointerSegment(path) + "/" + method
			_, info, err := resolveSwagger20Operation(d, ref)
			if err == nil {
				result = append(result, info)
			}
		}
	}
	return result
}
